Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,701 @@
+    det = getAtXY(this->xpixels() - 1, 0);
+    det->getBoundingBox(compBox);
+    m_cachedBoundingBox->grow(compBox);
+    det = getAtXY(this->xpixels() - 1, this->ypixels() - 1);
+    det->getBoundingBox(compBox);
+    m_cachedBoundingBox->grow(compBox);
+    det = getAtXY(0, this->ypixels() - 1);
+    det->getBoundingBox(compBox);
+    m_cachedBoundingBox->grow(compBox);
+  }
+  // Use cached box
+  assemblyBox = *m_cachedBoundingBox;
+}
+
+/**
+* Draws the objcomponent, If the handler is not set then this function does
+* nothing.
+*/
+void StructuredDetector::draw() const {
+  if (Handle() == nullptr)
+    return;
+  // Render the ObjComponent and then render the object
+  Handle()->Render();
+}
+
+/**
+* Draws the Object
+*/
+void StructuredDetector::drawObject() const {}
+
+/**
+* Initializes the ObjComponent for rendering, this function should be called
+* before rendering.
+*/
+void StructuredDetector::initDraw() const {
+  if (Handle() == nullptr)
+    return;
+
+  Handle()->Initialize();
+}
+
+/// Returns the shape of the Object
+const boost::shared_ptr<const Object> StructuredDetector::shape() const {
+  // --- Create a hexahedral shape for your pixels ----
+  auto w = this->xpixels() + 1;
+  auto xlb = m_xvalues[0];
+  auto xlf = m_xvalues[w * m_ypixels];
+  auto xrf = m_xvalues[(w * m_ypixels) + m_xpixels];
+  auto xrb = m_xvalues[w];
+  auto ylb = m_yvalues[0];
+  auto ylf = m_yvalues[(w * m_ypixels)];
+  auto yrf = m_yvalues[(w * m_ypixels) + m_xpixels];
+  auto yrb = m_yvalues[w];
+
+  std::ostringstream xmlShapeStream;
+
+  xmlShapeStream << "<hexahedron id=\"detector-shape\" >"
+                 << "<left-back-bottom-point x=\"" << xlb << "\""
+                 << " y=\"" << ylb << "\""
+                 << " z=\"0\" />"
+                 << "<left-front-bottom-point x=\"" << xlf << "\""
+                 << " y=\"" << ylf << "\""
+                 << " z=\"0\" />"
+                 << "<right-front-bottom-point x=\"" << xrf << "\""
+                 << " y=\"" << yrf << "\""
+                 << " z=\"0\" />"
+                 << "<right-back-bottom-point x=\"" << xrb << "\""
+                 << " y=\"" << yrb << "\""
+                 << " z=\"0\" />"
+                 << "<left-back-top-point x=\"" << xlb << "\""
+                 << " y=\"" << ylb << "\""
+                 << " z=\"0.5\" />"
+                 << "<left-front-top-point x=\"" << xlf << "\""
+                 << " y=\"" << ylf << "\""
+                 << " z=\"0.5\" />"
+                 << "<right-front-top-point x=\"" << xrf << "\""
+                 << " y=\"" << yrf << "\""
+                 << " z=\"0.5\" />"
+                 << "<right-back-top-point x=\"" << xrb << "\""
+                 << " y=\"" << yrb << "\""
+                 << " z=\"0.5\" />"
+                 << "</hexahedron>";
+
+  std::string xmlHexahedralShape(xmlShapeStream.str());
+  Geometry::ShapeFactory shapeCreator;
+  boost::shared_ptr<Geometry::Object> hexahedralShape =
```

Comment:
You can eliminate the named temporary here and just return the shared_ptr directly